File: smalivm/src/main/java/org/cf/smalivm/opcode/ReturnOpFactory.java
package org.cf.smalivm.opcode;

import gnu.trove.map.TIntObjectMap;

import org.cf.smalivm.VirtualMachine;
import org.jf.dexlib2.builder.MethodLocation;
import org.jf.dexlib2.iface.instruction.formats.Instruction11x;

public class ReturnOpFactory implements OpFactory {

    @Override
    public Op create(MethodLocation location, TIntObjectMap<MethodLocation> addressToLocation, VirtualMachine vm) {
        Instruction11x instr = (Instruction11x) location.getInstruction();
        int register = instr.getRegisterA();

        return new ReturnOp(location, register);
    }

}

File: smalivm/src/main/java/org/cf/smalivm/opcode/ReturnOpFactory.kt
package org.cf.smalivm.opcode

import org.cf.smalivm.VirtualMachine
import org.jf.dexlib2.builder.MethodLocation
import org.jf.dexlib2.iface.instruction.formats.Instruction11x

class ReturnOpFactory : OpFactory {
    override fun create(location: MethodLocation, addressToLocation: Map<Int, MethodLocation>, vm: VirtualMachine): Op {
        val instr = location.instruction as Instruction11x
        val register = instr.registerA
        return ReturnOp(location, register)
    }
}
